### PR TITLE
MANT: refactor unravel_index for code repetition

### PR DESCRIPTION
**Pull request by mattip**
Replaces #6219 which opened with the comment
"Adding a little bit of extra accounting in the inner loop allows joining the two specialized branches into a single one. The performance loss, if any, is within the measurement error."

Additional comments were integrated into this PR, including a nice little compiler optimization for getting the quotient and remainder with a single divide (thanks [godbolt](https://godbolt.org/z/tP-qwq) )

@seberg asked for a better error to be raised, which was already implemented.

I made a new PR since I could not push the rebase to @jaimefrio 's repo. Of course we only need one of these, @jaimefrio if you wish to move forward on it feel free to close this one.

### CODE REVIEW

File: numpy/core/src/multiarray/compiled_base.c [hunk 1]
```diff
@@ -1135,67 +1135,44 @@ arr_ravel_multi_index(PyObject *self, PyObject *args, PyObject *kwds)
...
-            *coords++ = val % unravel_dims[i];
-            val /= unravel_dims[i];
+            /* 
+             * Using a local seems to enable single-divide optimization
+             * but only if the / precedes the %
+             */
+            npy_intp tmp = val / unravel_dims[idx];
+            coords[idx] = val % unravel_dims[idx];
+            val = tmp;
+            idx += idx_step;
```
Comment (mattip) [inline numpy/core/src/multiarray/compiled_base.c]: This was the bit @charris pointed out could be [optimized](https://godbolt.org/z/tP-qwq) nicely by putting the divide first and using a temporary variable

File: numpy/core/src/multiarray/compiled_base.c [hunk 2]
```diff
@@ -1135,67 +1135,44 @@ arr_ravel_multi_index(PyObject *self, PyObject *args, PyObject *kwds)
...
+    int i, idx;
+    int idx_start = (order == NPY_CORDER) ? unravel_ndim - 1: 0;
+    int idx_step = (order == NPY_CORDER) ? -1 : 1;
+    char invalid = 0;
+    npy_intp val = 0;
 
     NPY_BEGIN_ALLOW_THREADS;
-    invalid = 0;
+    /* NPY_KEEPORDER or NPY_ANYORDER have no meaning in this setting */
+    assert (order == NPY_CORDER || order == NPY_FORTRANORDER);
```
Comment (seberg) [inline numpy/core/src/multiarray/compiled_base.c]:
```suggestion
    assert(order == NPY_CORDER || order == NPY_FORTRANORDER);
```

**Review by seberg — APPROVED** [review]
The diff looks a big, but the actual changes really seem very small. Just checked and the tests seem OK here, can only find some style nitpicks right now.

Would merge in a bit, unless someone wants to have another glance or wants me to look closer again.

**Comment by mattip** [comment]
I used the "accept suggested changes" feature but now the PR has 3 commits, some with strange messages. Probably need to squash-merge

**Comment by seberg** [comment]
Hmmm, so the feature is not perfect ;) (but there is not much it can do I guess, except possibly auto-squash multiple suggestions)... Will have another quick pass after lunch and wait a bit, but Jaime de-facto reviewed this as well I guess.

**Comment by seberg** [comment]
Thanks Matti, and of course also Jaime!

**Pull request merged**